File: ios/FprKit/Sources/FprKit/OOXMLAdapter.swift
import CryptoKit
import Foundation

/// Office Open XML documents encrypted with ECMA-376 *agile* encryption
/// ([MS-OFFCRYPTO] 2.3.4.10): an AES-encrypted package inside a CFB container,
/// with the key derived by an iterated SHA-512 over the password.
public struct OOXMLAdapter: Sendable {
    public init() {}

    // [MS-OFFCRYPTO] 2.3.4.12 -- block keys that diversify the derived key.
    private static let verifierInputBlockKey = Data([0xFE, 0xA7, 0xD2, 0x76, 0x3B, 0x4B, 0x9E, 0x79])
    private static let verifierValueBlockKey = Data([0xD7, 0xAA, 0x0F, 0x6D, 0x30, 0x61, 0x34, 0x4E])
    private static let keyValueBlockKey = Data([0x14, 0x6E, 0x0B, 0xE7, 0xAB, 0xAC, 0xD0, 0xD6])
    private static let segmentLength = 4096

    struct AgileDescriptor {
        var keyDataSalt = Data()
        var keyDataBlockSize = 16
        var spinCount = 100_000
        var passwordSalt = Data()
        var keyBits = 256
        var encryptedVerifierHashInput = Data()
        var encryptedVerifierHashValue = Data()
        var encryptedKeyValue = Data()
        var hashAlgorithm = "SHA512"
        var cipherAlgorithm = "AES"
    }

    public func detect(_ data: Data) throws -> Detection {
        if CFBReader.isCFB(data) {
            let container = try CFBReader(data)
            guard container.hasStream(named: "EncryptionInfo") else {
                return Detection(
                    format: .ooxml, protection: .unknown, removability: .unsupported,
                    detail: "Compound file without an EncryptionInfo stream."
                )
            }
            let descriptor = try readDescriptor(container)
            return Detection(
                format: .ooxml, protection: .userPassword, removability: .removable,
                algorithm:
                    "AES-\(descriptor.keyBits) (ECMA-376 agile, \(descriptor.hashAlgorithm), "
                    + "\(descriptor.spinCount) spins)",
                detail: "Encrypted. Supply the password the document was protected with."
            )
        }

        // An unencrypted OOXML package is an ordinary ZIP.
        let members = try ZipArchiveReader.readMembers(data)
        guard members["[Content_Types].xml"] != nil else {
            return Detection(
                format: .unknown, protection: .unknown, removability: .unsupported,
                detail: "Not an Office Open XML package."
            )
        }
        for (name, contents) in members where name.hasSuffix("settings.xml") {
            guard let text = String(data: contents, encoding: .utf8) else { continue }
            if text.contains("documentProtection") {
                return Detection(
                    format: .ooxml, protection: .ownerRestrictions, removability: .refused,
                    detail: """
                        Marked read-only with documentProtection. The content is not \
                        encrypted, so removing that flag is a bypass rather than \
                        decryption -- this tool does not do it.
                        """
                )
            }
        }
        return Detection(
            format: .ooxml, protection: .none, removability: .notProtected,
            detail: "Not encrypted. \(members.count) part(s)."
        )
    }

    public func remove(_ data: Data, password: String) throws -> Data {
        guard CFBReader.isCFB(data) else {
            let detection = try detect(data)
            throw FprError.policyRefused(
                detection.protection == .ownerRestrictions
                    ? detection.detail
                    : "This document is not encrypted; there is nothing to remove."
            )
        }
        let container = try CFBReader(data)
        let descriptor = try readDescriptor(container)

        let base = iteratedHash(password: password, salt: descriptor.passwordSalt, spinCount: descriptor.spinCount)
        let keyLength = descriptor.keyBits / 8

        // Verify before doing anything else: a wrong password stops here.
        let verifierInputKey = derive(base, Self.verifierInputBlockKey, length: keyLength)
        let verifierValueKey = derive(base, Self.verifierValueBlockKey, length: keyLength)
        let verifierInput = try Crypto.aesCBCDecrypt(
            descriptor.encryptedVerifierHashInput, key: verifierInputKey, iv: descriptor.passwordSalt
        )
        let expectedHash = try Crypto.aesCBCDecrypt(
            descriptor.encryptedVerifierHashValue, key: verifierValueKey, iv: descriptor.passwordSalt
        )
        let actualHash = Crypto.sha512(verifierInput.prefix(16))
        guard Crypto.constantTimeEquals(
            Data(actualHash.prefix(64)), Data(expectedHash.prefix(64))
        ) else {
            throw FprError.wrongPassword
        }

        let keyValueKey = derive(base, Self.keyValueBlockKey, length: keyLength)
        let secretKey = try Crypto.aesCBCDecrypt(
            descriptor.encryptedKeyValue, key: keyValueKey, iv: descriptor.passwordSalt
        ).prefix(keyLength)

        let encryptedPackage = try container.stream(named: "EncryptedPackage")
        return try decryptPackage(
            encryptedPackage, secretKey: Data(secretKey), keyDataSalt: descriptor.keyDataSalt,
            blockSize: descriptor.keyDataBlockSize
        )
    }

    private func decryptPackage(
        _ package: Data, secretKey: Data, keyDataSalt: Data, blockSize: Int
    ) throws -> Data {
        guard package.count >= 8 else {
            throw FprError.corruptFile("EncryptedPackage stream is too short")
        }
        var reader = ByteReader(package)
        let declaredLength = Int(try reader.u64())
        var output = Data()
        output.reserveCapacity(declaredLength)

        var segment: UInt32 = 0
        while reader.remaining > 0 {
            let take = min(Self.segmentLength, reader.remaining)
            // Each segment gets its own IV, derived from the salt and its index.
            var salted = keyDataSalt
            salted.append(contentsOf: [
                UInt8(segment & 0xFF), UInt8((segment >> 8) & 0xFF),
                UInt8((segment >> 16) & 0xFF), UInt8((segment >> 24) & 0xFF),
            ])
            let iv = Crypto.sha512(salted).prefix(blockSize)
            let chunk = try reader.bytes(take - (take % 16))
            if take % 16 != 0 { _ = try? reader.bytes(take % 16) }
            output.append(try Crypto.aesCBCDecrypt(chunk, key: secretKey, iv: Data(iv)))
            segment += 1
        }

        guard declaredLength <= output.count else {
            throw FprError.corruptFile(
                "declared package length \(declaredLength) exceeds \(output.count) decrypted bytes"
            )
        }
        return output.prefix(declaredLength)
    }

    /// H0 = SHA512(salt || UTF16LE(password)); Hn = SHA512(LE32(n) || Hn-1).
    private func iteratedHash(password: String, salt: Data, spinCount: Int) -> Data {
        var hash = Crypto.sha512(salt + Data(Array(password.utf16).flatMap {
            [UInt8($0 & 0xFF), UInt8($0 >> 8)]
        }))
        for index in 0..<spinCount {
            var block = Data([
                UInt8(index & 0xFF), UInt8((index >> 8) & 0xFF),
                UInt8((index >> 16) & 0xFF), UInt8((index >> 24) & 0xFF),
            ])
            block.append(hash)
            hash = Crypto.sha512(block)
        }
        return hash
    }

    private func derive(_ hash: Data, _ blockKey: Data, length: Int) -> Data {
        Data(Crypto.sha512(hash + blockKey).prefix(length))
    }

    private func readDescriptor(_ container: CFBReader) throws -> AgileDescriptor {
        let stream = try container.stream(named: "EncryptionInfo")
        guard stream.count > 8 else {
            throw FprError.corruptFile("EncryptionInfo stream is too short")
        }
        var header = ByteReader(stream)
        let major = try header.u16()
        let minor = try header.u16()
        guard major == 4, minor == 4 else {
            throw FprError.unsupportedFormat(
                """
                This document uses ECMA-376 \(major).\(minor) encryption. Only agile \
                encryption (4.4) is supported.
                """
            )
        }
        _ = try header.u32()  // reserved flags
        let xml = stream.dropFirst(8)

        let parser = AgileDescriptorParser()
        guard let descriptor = parser.parse(Data(xml)) else {
            throw FprError.corruptFile("EncryptionInfo descriptor could not be parsed")
        }
        return descriptor
    }
}

/// Pulls the two attribute-only elements out of the agile descriptor.
private final class AgileDescriptorParser: NSObject, XMLParserDelegate {
    private var descriptor = OOXMLAdapter.AgileDescriptor()
    private var sawKeyData = false
    private var sawEncryptedKey = false

    func parse(_ data: Data) -> OOXMLAdapter.AgileDescriptor? {
        let parser = XMLParser(data: data)
        parser.delegate = self
        guard parser.parse(), sawKeyData, sawEncryptedKey else { return nil }
        return descriptor
    }

    func parser(
        _ parser: XMLParser, didStartElement elementName: String, namespaceURI: String?,
        qualifiedName: String?, attributes: [String: String]
    ) {
        let local = elementName.split(separator: ":").last.map(String.init) ?? elementName
        func base64(_ key: String) -> Data {
            attributes[key].flatMap { Data(base64Encoded: $0) } ?? Data()
        }
        switch local {
        case "keyData":
            sawKeyData = true
            descriptor.keyDataSalt = base64("saltValue")
            descriptor.keyDataBlockSize = attributes["blockSize"].flatMap(Int.init) ?? 16
            descriptor.cipherAlgorithm = attributes["cipherAlgorithm"] ?? "AES"
        case "encryptedKey":
            sawEncryptedKey = true
            descriptor.spinCount = attributes["spinCount"].flatMap(Int.init) ?? 100_000
            descriptor.passwordSalt = base64("saltValue")
            descriptor.keyBits = attributes["keyBits"].flatMap(Int.init) ?? 256
            descriptor.hashAlgorithm = attributes["hashAlgorithm"] ?? "SHA512"
            descriptor.encryptedVerifierHashInput = base64("encryptedVerifierHashInput")
            descriptor.encryptedVerifierHashValue = base64("encryptedVerifierHashValue")
            descriptor.encryptedKeyValue = base64("encryptedKeyValue")
        default:
            break
        }
    }
}

File: android/fprkit/src/main/kotlin/dev/jayanth/fpr/OoxmlAdapter.kt
package dev.jayanth.fpr

import java.io.ByteArrayOutputStream
import java.util.Base64
import javax.xml.parsers.DocumentBuilderFactory
import org.w3c.dom.Element

/**
 * Office Open XML documents using ECMA-376 *agile* encryption
 * ([MS-OFFCRYPTO] 2.3.4.10): an AES-encrypted package inside a CFB container,
 * keyed by an iterated SHA-512 over the password.
 */
class OoxmlAdapter {

    // [MS-OFFCRYPTO] 2.3.4.12 -- block keys that diversify the derived key.
    private val verifierInputBlockKey =
        byteArrayOf(0xFE.toByte(), 0xA7.toByte(), 0xD2.toByte(), 0x76, 0x3B, 0x4B, 0x9E.toByte(), 0x79)
    private val verifierValueBlockKey =
        byteArrayOf(0xD7.toByte(), 0xAA.toByte(), 0x0F, 0x6D, 0x30, 0x61, 0x34, 0x4E)
    private val keyValueBlockKey =
        byteArrayOf(0x14, 0x6E, 0x0B, 0xE7.toByte(), 0xAB.toByte(), 0xAC.toByte(), 0xD0.toByte(), 0xD6.toByte())

    private data class Descriptor(
        val keyDataSalt: ByteArray,
        val keyDataBlockSize: Int,
        val spinCount: Int,
        val passwordSalt: ByteArray,
        val keyBits: Int,
        val hashAlgorithm: String,
        val encryptedVerifierHashInput: ByteArray,
        val encryptedVerifierHashValue: ByteArray,
        val encryptedKeyValue: ByteArray,
    )

    fun detect(data: ByteArray): Detection {
        if (CfbReader.isCfb(data)) {
            val container = CfbReader(data)
            if (!container.hasStream("EncryptionInfo")) {
                return Detection(
                    FormatId.OOXML, Protection.UNKNOWN, Removability.UNSUPPORTED,
                    detail = "Compound file without an EncryptionInfo stream.",
                )
            }
            val descriptor = readDescriptor(container)
            return Detection(
                FormatId.OOXML, Protection.USER_PASSWORD, Removability.REMOVABLE,
                "AES-${descriptor.keyBits} (ECMA-376 agile, ${descriptor.hashAlgorithm}, " +
                    "${descriptor.spinCount} spins)",
                "Encrypted. Supply the password the document was protected with.",
            )
        }

        val members = ZipArchive.readMembers(data)
        if (members["[Content_Types].xml"] == null) {
            return Detection(
                FormatId.UNKNOWN, Protection.UNKNOWN, Removability.UNSUPPORTED,
                detail = "Not an Office Open XML package.",
            )
        }
        for ((name, contents) in members) {
            if (!name.endsWith("settings.xml")) continue
            if (String(contents, Charsets.UTF_8).contains("documentProtection")) {
                return Detection(
                    FormatId.OOXML, Protection.OWNER_RESTRICTIONS, Removability.REFUSED,
                    detail = "Marked read-only with documentProtection. The content is not " +
                        "encrypted, so removing that flag is a bypass rather than decryption -- " +
                        "this tool does not do it.",
                )
            }
        }
        return Detection(
            FormatId.OOXML, Protection.NONE, Removability.NOT_PROTECTED,
            detail = "Not encrypted. ${members.size} part(s).",
        )
    }

    fun remove(data: ByteArray, password: String): ByteArray {
        if (!CfbReader.isCfb(data)) {
            val detection = detect(data)
            throw FprException.PolicyRefused(
                if (detection.protection == Protection.OWNER_RESTRICTIONS) detection.detail
                else "This document is not encrypted; there is nothing to remove."
            )
        }
        val container = CfbReader(data)
        val descriptor = readDescriptor(container)
        val base = iteratedHash(password, descriptor.passwordSalt, descriptor.spinCount)
        val keyLength = descriptor.keyBits / 8

        // Verify before anything else: a wrong password stops here.
        val verifierInput = Crypto.aesCbcDecrypt(
            descriptor.encryptedVerifierHashInput,
            derive(base, verifierInputBlockKey, keyLength), descriptor.passwordSalt,
        )
        val expectedHash = Crypto.aesCbcDecrypt(
            descriptor.encryptedVerifierHashValue,
            derive(base, verifierValueBlockKey, keyLength), descriptor.passwordSalt,
        )
        val actualHash = Crypto.sha512(verifierInput.copyOfRange(0, 16))
        if (!Crypto.constantTimeEquals(actualHash, expectedHash.copyOfRange(0, 64))) {
            throw FprException.WrongPassword()
        }

        val secretKey = Crypto.aesCbcDecrypt(
            descriptor.encryptedKeyValue, derive(base, keyValueBlockKey, keyLength),
            descriptor.passwordSalt,
        ).copyOfRange(0, keyLength)

        return decryptPackage(
            container.stream("EncryptedPackage"), secretKey, descriptor.keyDataSalt,
            descriptor.keyDataBlockSize,
        )
    }

    private fun decryptPackage(
        packageStream: ByteArray, secretKey: ByteArray, keyDataSalt: ByteArray, blockSize: Int,
    ): ByteArray {
        if (packageStream.size < 8) {
            throw FprException.CorruptFile("EncryptedPackage stream is too short")
        }
        val reader = ByteReader(packageStream)
        val declaredLength = reader.u64().toInt()
        val output = ByteArrayOutputStream(maxOf(declaredLength, 0))

        var segment = 0
        while (reader.remaining > 0) {
            val take = minOf(SEGMENT_LENGTH, reader.remaining)
            // Each segment gets its own IV, derived from the salt and its index.
            val salted = keyDataSalt + byteArrayOf(
                (segment and 0xFF).toByte(), ((segment shr 8) and 0xFF).toByte(),
                ((segment shr 16) and 0xFF).toByte(), ((segment shr 24) and 0xFF).toByte(),
            )
            val iv = Crypto.sha512(salted).copyOfRange(0, blockSize)
            val whole = take - (take % 16)
            val chunk = reader.bytes(whole)
            if (take % 16 != 0) reader.bytes(take % 16)
            output.write(Crypto.aesCbcDecrypt(chunk, secretKey, iv))
            segment += 1
        }

        val decrypted = output.toByteArray()
        if (declaredLength > decrypted.size) {
            throw FprException.CorruptFile(
                "declared package length $declaredLength exceeds ${decrypted.size} decrypted bytes"
            )
        }
        return decrypted.copyOf(declaredLength)
    }

    /** H0 = SHA512(salt || UTF16LE(password)); Hn = SHA512(LE32(n) || Hn-1). */
    private fun iteratedHash(password: String, salt: ByteArray, spinCount: Int): ByteArray {
        var hash = Crypto.sha512(salt + password.toByteArray(Charsets.UTF_16LE))
        for (index in 0 until spinCount) {
            val block = byteArrayOf(
                (index and 0xFF).toByte(), ((index shr 8) and 0xFF).toByte(),
                ((index shr 16) and 0xFF).toByte(), ((index shr 24) and 0xFF).toByte(),
            )
            hash = Crypto.sha512(block + hash)
        }
        return hash
    }

    private fun derive(hash: ByteArray, blockKey: ByteArray, length: Int): ByteArray =
        Crypto.sha512(hash + blockKey).copyOfRange(0, length)

    private fun readDescriptor(container: CfbReader): Descriptor {
        val stream = container.stream("EncryptionInfo")
        if (stream.size <= 8) throw FprException.CorruptFile("EncryptionInfo stream is too short")
        val header = ByteReader(stream)
        val major = header.u16()
        val minor = header.u16()
        if (major != 4 || minor != 4) {
            throw FprException.UnsupportedFormat(
                "This document uses ECMA-376 $major.$minor encryption. Only agile " +
                    "encryption (4.4) is supported."
            )
        }

        val xml = stream.copyOfRange(8, stream.size)
        // This XML comes from the file being inspected, so external entities
        // must never be resolved. Android's parser does not implement all of
        // these feature names -- it rejects them outright rather than ignoring
        // them -- so each is set independently and the EntityResolver below is
        // the backstop that works on every runtime.
        val factory = DocumentBuilderFactory.newInstance()
        factory.isNamespaceAware = true
        for (feature in listOf(
            "http://apache.org/xml/features/disallow-doctype-decl",
            "http://xml.org/sax/features/external-general-entities",
            "http://xml.org/sax/features/external-parameter-entities",
            "http://apache.org/xml/features/nonvalidating/load-external-dtd",
        )) {
            runCatching {
                factory.setFeature(feature, feature.endsWith("disallow-doctype-decl"))
            }
        }
        runCatching { factory.isXIncludeAware = false }
        factory.isExpandEntityReferences = false

        val builder = factory.newDocumentBuilder()
        builder.setEntityResolver { _, _ ->
            org.xml.sax.InputSource(java.io.ByteArrayInputStream(ByteArray(0)))
        }
        val document = builder.parse(xml.inputStream())

        fun element(local: String): Element = document.getElementsByTagName("*").let { nodes ->
            (0 until nodes.length).asSequence()
                .mapNotNull { nodes.item(it) as? Element }
                .firstOrNull { it.tagName.substringAfterLast(':') == local }
                ?: throw FprException.CorruptFile("<$local> missing from EncryptionInfo")
        }

        val keyData = element("keyData")
        val encryptedKey = element("encryptedKey")
        fun Element.base64(name: String): ByteArray =
            Base64.getDecoder().decode(getAttribute(name).ifEmpty { "" })

        return Descriptor(
            keyDataSalt = keyData.base64("saltValue"),
            keyDataBlockSize = keyData.getAttribute("blockSize").toIntOrNull() ?: 16,
            spinCount = encryptedKey.getAttribute("spinCount").toIntOrNull() ?: 100_000,
            passwordSalt = encryptedKey.base64("saltValue"),
            keyBits = encryptedKey.getAttribute("keyBits").toIntOrNull() ?: 256,
            hashAlgorithm = encryptedKey.getAttribute("hashAlgorithm").ifEmpty { "SHA512" },
            encryptedVerifierHashInput = encryptedKey.base64("encryptedVerifierHashInput"),
            encryptedVerifierHashValue = encryptedKey.base64("encryptedVerifierHashValue"),
            encryptedKeyValue = encryptedKey.base64("encryptedKeyValue"),
        )
    }

    private companion object {
        const val SEGMENT_LENGTH = 4096
    }
}
